src/java.base/share/classes/java/io/SerializationMisdeclarationChecker.java

```diff
@@ -0,0 +1,300 @@
+    private static final String SERIAL_PERSISTENT_FIELDS_NAME = "serialPersistentFields";
+    private static final String WRITE_OBJECT_NAME = "writeObject";
+    private static final String READ_OBJECT_NAME = "readObject";
+    private static final String READ_OBJECT_NO_DATA_NAME = "readObjectNoData";
+    private static final String WRITE_REPLACE_NAME = "writeReplace";
+    private static final String READ_RESOLVE_NAME = "readResolve";
+
+    private static final Class<?>[] WRITE_OBJECT_PARAM_TYPES = {ObjectOutputStream.class};
+    private static final Class<?>[] READ_OBJECT_PARAM_TYPES = {ObjectInputStream.class};
+    private static final Class<?>[] READ_OBJECT_NO_DATA_PARAM_TYPES = {};
+    private static final Class<?>[] WRITE_REPLACE_PARAM_TYPES = {};
+    private static final Class<?>[] READ_RESOLVE_PARAM_TYPES = {};
+
+    private final Class<?> cl;
+
+    SerializationMisdeclarationChecker(Class<?> cl) {
+        this.cl = cl;
+    }
+
+    @SuppressWarnings("removal")
+    void checkMisdeclarations() {
+        AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
+            privilegedCheckSerialVersionUID();
+            privilegedCheckSerialPersistentFields();
+
+            privilegedCheckPrivateMethod(WRITE_OBJECT_NAME, WRITE_OBJECT_PARAM_TYPES, Void.TYPE);
+            privilegedCheckPrivateMethod(READ_OBJECT_NAME, READ_OBJECT_PARAM_TYPES, Void.TYPE);
+            privilegedCheckPrivateMethod(READ_OBJECT_NO_DATA_NAME, READ_OBJECT_NO_DATA_PARAM_TYPES, Void.TYPE);
+
+            privilegedCheckAccessibleMethod(WRITE_REPLACE_NAME, WRITE_REPLACE_PARAM_TYPES, Object.class);
+            privilegedCheckAccessibleMethod(READ_RESOLVE_NAME, READ_RESOLVE_PARAM_TYPES, Object.class);
+
+            return null;
+        });
+    }
+
+    private void privilegedCheckSerialVersionUID() {
+        Field f = declaredField(cl, SUID_NAME);
+        if (f == null) {
+            if (isOrdinaryClass()) {
+                commitEvent(SUID_EXPLICIT,
+                        "a private static final long " + SUID_NAME +
+                                " field should be declared");
+            }
+            return;
+        }
+        if (cl.isEnum()) {
+            commitEvent(SUID_INEFFECTIVE_ENUM,
+                    SUID_NAME + " is not effective in an enum class");
+        }
+        if (!isPrivate(f)) {
+            commitEvent(SUID_PRIVATE,
+                    SUID_NAME + " should be declared private");
+        }
+        if (!isStatic(f)) {
+            commitEvent(SUID_STATIC,
+                    SUID_NAME + " must be declared static to be effective");
+        }
+        if (!isFinal(f)) {
+            commitEvent(SUID_FINAL,
+                    SUID_NAME + " must be declared final to be effective");
+        }
+        if (f.getType() != Long.TYPE) {
+            commitEvent(SUID_LONG,
+                    SUID_NAME + " should be declared of type long");
+        }
+        if (!isStatic(f)) {
```

Comment:
Nit - perhaps save the return value from the previous call to `isStatic(f)` a few lines above and reuse it here?

Reply:
Assuming that most serializable classes are declared correctly, I don't think that caching the value of `isStatic()` makes a measurable difference.
If at all, then I'd prefer to cache the modifiers and "inline" the masking logic of the individual is*() methods.